accepted/2021/preview-features/preview-features.md

```diff
@@ -0,0 +1,418 @@
+they are trying to figure out how they can integrate machine learning with
+ML.NET more tightly with their services, but they struggle with data ingestion
+because a lot of the heavy computations are happening inside of an old C++ base.
+When that code was written, .NET wasn't fast enough to do heavy numerical
+computations.
+
+As part of their research into this problem space, they notice a recent blog
+post about how the upcoming generic numeric support and hardware intrinsics can
+help to speed things up considerably. They decide to give it a spin, so they
+copy & paste some sample code from the blog post.
+
+The code immediately produces an error message:
+
+> error: The type 'IArithmetic\<T>' is in preview. In order to use it, you need
+> to enable preview features.
+
+After taking a quick look at the blog post, Ainsley realizes that they only
+needs to set `EnablePreviewFeatures` to `true` in the project file.
+
+### Consuming a library that uses a preview feature
+
+Sahibi works with Ainsley and tries to consume their library that exposes some
+of the core algorithms so she can integrate those with ML.NET. After adding the
+reference to Ainsley's library `Fabrikam.Algorithms.Core` she instantiates the
+engine type which produces the following error:
+
+> The library 'Fabrikam.Algorithms.Core' uses preview features of the .NET
+> platform. In order to use it, you need to enable preview features.
+
+Sahibi IMs Ainsley to ask what this is about where she learns that this feature
+isn't a supported part of .NET, therefore she should make sure to not integrate
+the library into the production instance of Fabrikam.
+
+## Requirements
+
+### Goals
+
+* Customers can discover preview features
+* Customers are made aware if a particular feature is in a preview state so that
+  they always know which part of the product is fully supported and which parts
+  aren't.
+* Customer have an easy way to audit if their applications or libraries are
+  using preview features.
+* Customers can't accidentally depend on preview features, either directly or
+  indirectly through some library.
+* We can deliver a reasonable experience even when a preview feature can't be
+  supported in all the workloads, for example, by failing with a sensible error
+  message.
+* We have a unified experience for enabling preview features, no matter whether
+  they are in a runtime, a compiler, a library, or span multiple parts of the
+  system.
+* We don't have to do unnatural acts to expose a preview feature, by, for
+  example, having to build a plug-in based system that allows swapping in and
+  out files.
+
+### Non-Goals
+
+* Allowing Microsoft to turn features on or off remotely, e.g. A/B testing
+* Designing a mechanism for shipping preview features as separate packages or
+  installers
+
+## Stakeholders and Reviewers
+
+* Runtime teams
+* C#, VB, F# compiler and IDE teams
+* Library teams
+* Project system team
+* SDK team
+* MSBuild team
+* NuGet team
```

Comment:
Should Powershell be also in this list? They may need to do work to prevent powershell scripts taking accidental depedency on unstable preview surface.

Reply:
I'll add them